toolbox/src/DiscreteFilter.cpp

```diff
@@ -0,0 +1,317 @@
+    }
+    // MedianFilter
+    // ------------
+    else if (filter_type == "MedianFilter") {
+        if (medianFilter_order.int32Data() == 0) {
+            if (error) {
+                error->message = ClassName
+                                 + " (MedianFilter) You need to specify the "
+                                   "filter order.";
+            }
+            return false;
+        }
+        filter = new MedianFilter(medianFilter_order.int32Data());
+    }
+    else {
+        if (error) error->message = ClassName + " Filter type not recognized.";
+        return false;
+    }
+
+    // Initialize the other data
+    // =========================
+
+    // Get the width of the input vector
+    inputSignalWidth = blockInfo->getInputPortWidth(INPUT_IDX_SIGNAL);
+
+    // Check the initial conditions are properly sized
+    unsigned outputSignalWidth = blockInfo->getInputPortWidth(OUTPUT_IDX_SIGNAL);
+    unsigned y0Width           = y0->length();
+    unsigned u0Width           = u0->length();
+
+    if ((y0_str != "none") && (y0Width != outputSignalWidth)) {
+        if (error) {
+            error->message = ClassName + " y0 and output signal sizes don't match";
+        }
+        return false;
+    }
+
+    if ((u0_str != "none") && (filter_type == "Generic") && (u0Width != inputSignalWidth)) {
+        if (error) {
+            error->message = ClassName + " (Generic) u0 and input signal sizes don't match";
+        }
+        return false;
+    }
+
+    // Allocate the input signal
+    inputSignalVector = new Vector(inputSignalWidth, 0.0);
+
+    return true;
+}
+
+bool DiscreteFilter::terminate(BlockInformation* blockInfo, wbt::Error* error)
+{
+    // Deallocate all the memory
+    // -------------------------
+
+    if (filter) {
+        delete filter;
+        filter = nullptr;
+    }
+
+    if (inputSignalVector) {
+        delete inputSignalVector;
+        inputSignalVector = nullptr;
+    }
+
+    if (y0) {
+        delete y0;
+        y0 = nullptr;
+    }
+
+    if (u0) {
+        delete u0;
+        u0 = nullptr;
+    }
+
+    return true;
+}
+
+bool DiscreteFilter::initializeInitialConditions(BlockInformation* blockInfo, wbt::Error* error)
+{
+    // Reminder: this function is called when, during runtime, a block is disabled
+    // and enabled again. The method ::initialize instead is called just one time.
+
+    // If the initial conditions for the output are not set, allocate a properly
+    // sized vector
+    if (*y0 == Vector(0)) {
+        y0 = new Vector(inputSignalWidth, 0.0);
+    }
+
+    // Initialize the filter. This is required because if the signal is not 1D,
+    // the default filter constructor initialize a wrongly sized y0
+    // Moreover, the Filter class has a different constructor that handles the
+    // zero-gain case
+    Filter* filter_c = dynamic_cast<Filter*>(filter);
+    if (filter_c != nullptr) {
+        filter_c->init(*y0, *u0);
```

Comment:
I don't remember what `init` does, but is it ok to call with an empty vector? (i.e. `u0.size() == 0`)